lib/FtpConnection.js

```diff
@@ -1073,14 +1073,31 @@ FtpConnection.prototype._command_RNFR = function(commandArg) {
 FtpConnection.prototype._command_RNTO = function(commandArg) {
   var self = this;
   var fileto = withCwd(self.cwd, commandArg);
-  self.fs.rename(pathModule.join(self.root, self.filefrom), pathModule.join(self.root, fileto), function(err) {
-    if (err) {
-      self._logIf(LOG.ERROR, 'Error renaming file from ' + self.filefrom + ' to ' + fileto);
-      self.respond('550 Rename failed' + (err.code === 'ENOENT' ? '; file does not exist' : ''));
-    } else {
-      self.respond('250 File renamed successfully');
-    }
-  });
+
+  // Filter extensions if whitelist available
+  var whitelist = self.server.options.allowedExtensions;
+  var accepted = true;
+  if (whitelist && typeof whitelist === 'object' && Array.isArray(whitelist) && whitelist.length) {
+    accepted = false;
+    whitelist.forEach(function (ext) {
+      if (fileto.endsWith(ext)) accepted = true;
```

Comment:
We should allow RegExps too, or Cartesian product will explode for relatively simple stuff like `/\.(html?|jpe?g|png|gif|ppm|txt)(\.(t?gz|zip|tar)|)$/`. And that doesn't even yet include rules to allow some parts to be all-uppercase when using tar. ;-)